src/librustc/middle/dead.rs

```diff
@@ -189,6 +189,24 @@ impl<'a, 'tcx> MarkSymbolVisitor<'a, 'tcx> {
         self.struct_has_extern_repr = had_extern_repr;
         self.inherited_pub_visibility = had_inherited_pub_visibility;
     }
+
+    fn mark_as_used_if_union(&mut self, did: DefId) {
+        if let Some(node_id) = self.tcx.hir.as_local_node_id(did) {
+            match self.tcx.hir.find(node_id) {
+                Some(hir_map::NodeItem(item)) => match item.node {
```

Comment:
Oh, sorry, I meant `hir::ExprStruct` in https://github.com/rust-lang/rust/pull/43397#discussion_r130736804 of course, not `hir::ExprPath`.
`mark_as_used_if_union` is good except that it should add only fields used in `ExprStruct` to the `live_symbols` set, not all fields from the union.
Nit: `mark_as_used_if_union` could also use some `if let`s.